endpoints/openrtb2/auction.go

```diff
@@ -1378,42 +1394,44 @@ func (deps *endpointDeps) processStoredRequests(ctx context.Context, requestJson
-	for i := 0; i < len(impIds); i++ {
-		resolvedImp, err := jsonpatch.MergePatch(storedImps[impIds[i]], imps[idIndices[i]])
+	impExtInfoMap := make(map[string]exchange.ImpExtInfo, 0)
+	resolvedImps := make([]json.RawMessage, 0)
+	for i, impData := range impInfo {
+		if impData.ImpExtPrebid.StoredRequest != nil && len(impData.ImpExtPrebid.StoredRequest.ID) > 0 {
+			resolvedImp, err := jsonpatch.MergePatch(storedImps[impData.ImpExtPrebid.StoredRequest.ID], impData.Imp)
 
-		if err != nil {
-			hasErr, Err := getJsonSyntaxError(imps[idIndices[i]])
-			if hasErr {
-				err = fmt.Errorf("Invalid JSON in Imp[%d] of Incoming Request: %s", i, Err)
-			} else {
-				hasErr, Err = getJsonSyntaxError(storedImps[impIds[i]])
+			if err != nil {
+				hasErr, Err := getJsonSyntaxError(impData.Imp)
 				if hasErr {
-					err = fmt.Errorf("imp.ext.prebid.storedrequest.id %s: Stored Imp has Invalid JSON: %s", impIds[i], Err)
+					err = fmt.Errorf("Invalid JSON in Imp[%d] of Incoming Request: %s", i, Err)
+				} else {
+					hasErr, Err = getJsonSyntaxError(storedImps[impData.ImpExtPrebid.StoredRequest.ID])
+					if hasErr {
+						err = fmt.Errorf("imp.ext.prebid.storedrequest.id %s: Stored Imp has Invalid JSON: %s", impData.ImpExtPrebid.StoredRequest.ID, Err)
+					}
 				}
+				return nil, nil, []error{err}
 			}
-			return nil, nil, []error{err}
-		}
-		imps[idIndices[i]] = resolvedImp
+			resolvedImps = append(resolvedImps, resolvedImp)
 
-		impId, err := jsonparser.GetString(resolvedImp, "id")
-		if err != nil {
-			return nil, nil, []error{err}
-		}
-		// This is substantially faster for reading values from a json blob than the Go json package,
-		// but keep in mind that each jsonparser.GetXXX call re-parses the entire json.
-		// Based on performance measurements, the tipping point of efficiency is around 4 calls.
-		// At that point, please consider switching to EachKey to use a single pass.
-		includeVideoAttributes, err := jsonparser.GetBoolean(resolvedImp, "ext", "prebid", "options", "echovideoattrs")
-		if err != nil && err != jsonparser.KeyPathNotFoundError {
-			return nil, nil, []error{err}
-		}
-		if storedImps[impIds[i]] != nil {
-			impExtInfoMap[impId] = exchange.ImpExtInfo{EchoVideoAttrs: includeVideoAttributes, StoredImp: storedImps[impIds[i]]}
+			impId, err := jsonparser.GetString(resolvedImp, "id")
+			if err != nil {
+				return nil, nil, []error{err}
+			}
+
+			echovideoAttributes := false
+			if impData.ImpExtPrebid.Options != nil {
+				echovideoAttributes = impData.ImpExtPrebid.Options.Echovideoattrs
+			}
+			impExtInfoMap[impId] = exchange.ImpExtInfo{EchoVideoAttrs: echovideoAttributes, StoredImp: storedImps[impData.ImpExtPrebid.StoredRequest.ID]}
+
+		} else {
+			resolvedImps = append(resolvedImps, impData.Imp)
 		}
+
 	}
-	if len(impIds) > 0 {
-		newImpJson, err := json.Marshal(imps)
+	if len(resolvedImps) > 0 {
+		newImpJson, err := json.Marshal(resolvedImps)
```

Comment:
I'll be honest that's it a bit hard for me to follow the changes in this function. Why is this now using `resolvedImps` when before it was using `imps`?

Reply:
Idea here is to simplify extraction of stored impression id. Before there were 3 structures in code to do this:
- `impIds` - is not imp ids, it's stored imp ids
- `imps` - impressions itself
- `idIndices` - specifies what imp index has stored impId, this is extra variable to support requests where not all imps have stored imps.

So we had to do this `imps[idIndices[i]] = resolvedImp` in order to find what impression had what id. 
Please refer to this comment: https://github.com/prebid/prebid-server/pull/1889#discussion_r681340201

Also part of this PR is a base for future stored responses implementation. 
Here we only have 

```
type ExtImpPrebid struct {
	// StoredRequest specifies which stored impression to use, if any.
	StoredRequest *ExtStoredRequest `json:"storedrequest"`

	// IsRewardedInventory is a signal intended for video impressions. Must be 0 or 1.
	IsRewardedInventory int8 `json:"is_rewarded_inventory"`

	// Bidder is the preferred approach for providing paramters to be interepreted by the bidder's adapter.
	Bidder map[string]json.RawMessage `json:"bidder"`

	Options *Options `json:"options,omitempty"`
}
``` 
which is used as a part of impression.ext parsing to this struct:
```
type ImpExtPrebidData struct {
	Imp          json.RawMessage
	ImpExtPrebid openrtb_ext.ExtImpPrebid
}
```
For Stored responses implementation ExtImpPrebid will be extended with 

```
// ExtStoredAuctionResponse specifies which stored auction response to use, if any.
ExtStoredAuctionResponse *ExtStoredAuctionResponse `json:"storedauctionresponse,omitempty"`

// ExtStoredBidResponse specifies which stored bid responseto use, if any.
ExtStoredBidResponse []*ExtStoredBidResponse `json:"storedbidresponse,omitempty"`
```

Hope this makes sense